contracts/access/AccessControlAdminRules.sol

```diff
@@ -0,0 +1,228 @@
+     */
+    function pendingAdmin() public view virtual returns (address) {
+        return _pendingAdmin;
+    }
+
+    /**
+     * @dev See {IERC165-supportsInterface}.
+     */
+    function supportsInterface(bytes4 interfaceId) public view virtual override returns (bool) {
+        return interfaceId == type(IAccessControlAdminRules).interfaceId || super.supportsInterface(interfaceId);
+    }
+
+    /**
+     * @dev See {IAccessControlAdminRules-beginAdminTransfer}
+     */
+    function beginAdminTransfer(address newAdmin) public virtual override onlyRole(DEFAULT_ADMIN_ROLE) {
+        require(delayedUntil() == 0, "AccessControl: pending admin already set");
+        _delayedUntil = SafeCast.toUint48(block.timestamp) + _delay;
+        _pendingAdmin = newAdmin;
+        emit AdminRoleChangeStarted(pendingAdmin(), delayedUntil());
+    }
+
+    /**
+     * @dev See {IAccessControlAdminRules-acceptAdminTransfer}
+     */
+    function acceptAdminTransfer() public virtual {
+        address pendingAdminOwner = pendingAdmin();
+        require(
+            _hasAdminTransferDelayPassed() && _msgSender() == pendingAdminOwner,
+            "AccessControl: delay must be met and caller must be pending admin"
+        );
+        _revokeRole(DEFAULT_ADMIN_ROLE, admin());
+        _grantRole(DEFAULT_ADMIN_ROLE, pendingAdminOwner);
+        _resetAdminTransfer();
+    }
+
+    /**
+     * @dev See {IAccessControlAdminRules-cancelAdminTransfer}
+     */
+    function cancelAdminTransfer() public virtual onlyRole(DEFAULT_ADMIN_ROLE) {
+        _resetAdminTransfer();
+    }
+
+    /**
+     * @dev Revokes `role` from the calling account.
+     *
+     * For `DEFAULT_ADMIN_ROLE`, only allows renouncing in two steps, so it's required
+     * that the {delayedUntil} is met and the pending admin is the zero address.
+     * After its execution, it will not be possible to call `onlyRole(DEFAULT_ADMIN_ROLE)`
+     * functions.
+     *
+     * For other roles, see {AccessControl-renounceRole}.
+     *
+     * NOTE: Renouncing `DEFAULT_ADMIN_ROLE` will leave the contract without an owner,
+     * thereby disabling any functionality that is only available to the default admin, and the
+     * possibility of reassigning a non-administrated role.
+     */
+    function renounceRole(bytes32 role, address account) public virtual override(IAccessControl, AccessControl) {
+        if (role == DEFAULT_ADMIN_ROLE) {
+            require(
+                pendingAdmin() == address(0) && _hasAdminTransferDelayPassed(),
+                "AccessControl: admin can only renounce in two delayed steps"
+            );
+        }
+        super.renounceRole(role, account);
+    }
+
+    /**
+     * @dev See {AccessControl-grantRole}. Reverts for `DEFAULT_ADMIN_ROLE`.
+     */
+    function grantRole(
+        bytes32 role,
+        address account
+    ) public virtual override(IAccessControl, AccessControl) onlyRole(getRoleAdmin(role)) {
```

Comment:
This check will already be done in `super.grantRole`.

```suggestion
    ) public virtual override(IAccessControl, AccessControl) {
```

Reply:
Tested, it works. Thanks!